File: apps/elixir/design-pattern/lib/design_pattern/interpreter.ex
defmodule DesignPattern.Interpreter do
  @moduledoc """
  Interpreter パターン: タグ付きタプルに対する再帰的パターンマッチングで
  DSL（ドメイン固有言語）を解釈・評価する。
  """

  @doc "数値リテラル"
  def literal(value), do: {:literal, value}

  @doc "加算式"
  def add(left, right), do: {:add, left, right}

  @doc "減算式"
  def subtract(left, right), do: {:subtract, left, right}

  @doc "乗算式"
  def multiply(left, right), do: {:multiply, left, right}

  @doc "除算式"
  def divide(left, right), do: {:divide, left, right}

  @doc "変数参照"
  def variable(name), do: {:variable, name}

  @doc "式を評価する"
  def evaluate(expr, env \\ %{})

  def evaluate({:literal, value}, _env), do: value

  def evaluate({:variable, name}, env) do
    Map.fetch!(env, name)
  end

  def evaluate({:add, left, right}, env) do
    evaluate(left, env) + evaluate(right, env)
  end

  def evaluate({:subtract, left, right}, env) do
    evaluate(left, env) - evaluate(right, env)
  end

  def evaluate({:multiply, left, right}, env) do
    evaluate(left, env) * evaluate(right, env)
  end

  def evaluate({:divide, left, right}, env) do
    divisor = evaluate(right, env)

    if divisor == 0 do
      {:error, :division_by_zero}
    else
      evaluate(left, env) / divisor
    end
  end

  @doc "式を文字列に変換する"
  def to_string_expr({:literal, value}), do: "#{value}"
  def to_string_expr({:variable, name}), do: "#{name}"

  def to_string_expr({:add, left, right}) do
    "(#{to_string_expr(left)} + #{to_string_expr(right)})"
  end

  def to_string_expr({:subtract, left, right}) do
    "(#{to_string_expr(left)} - #{to_string_expr(right)})"
  end

  def to_string_expr({:multiply, left, right}) do
    "(#{to_string_expr(left)} * #{to_string_expr(right)})"
  end

  def to_string_expr({:divide, left, right}) do
    "(#{to_string_expr(left)} / #{to_string_expr(right)})"
  end
end

File: apps/clojure/design-pattern/src/design_pattern/interpreter.clj
(ns design-pattern.interpreter
  (:require [clojure.string :as str]))

;; Interpreter パターン
;; AST をマップで表現し、マルチメソッドで evaluate する。

;; --- AST ノード生成ヘルパー ---

(defn literal [value]
  {:type :literal :value value})

(defn variable [name]
  {:type :variable :name name})

(defn add [left right]
  {:type :add :left left :right right})

(defn subtract [left right]
  {:type :subtract :left left :right right})

(defn multiply [left right]
  {:type :multiply :left left :right right})

(defn divide [left right]
  {:type :divide :left left :right right})

;; --- 評価 ---

(defmulti evaluate
  "AST ノードを環境 (変数マップ) の下で評価する。"
  (fn [node _env] (:type node)))

(defmethod evaluate :literal [node _env]
  (:value node))

(defmethod evaluate :variable [node env]
  (if-let [val (get env (:name node))]
    val
    (throw (ex-info (str "Undefined variable: " (:name node))
                    {:variable (:name node) :env env}))))

(defmethod evaluate :add [node env]
  (+ (evaluate (:left node) env)
     (evaluate (:right node) env)))

(defmethod evaluate :subtract [node env]
  (- (evaluate (:left node) env)
     (evaluate (:right node) env)))

(defmethod evaluate :multiply [node env]
  (* (evaluate (:left node) env)
     (evaluate (:right node) env)))

(defmethod evaluate :divide [node env]
  (let [divisor (evaluate (:right node) env)]
    (if (zero? divisor)
      (throw (ex-info "Division by zero" {:node node}))
      (/ (evaluate (:left node) env) divisor))))

;; --- 式の文字列表現 ---

(defmulti to-string
  "AST ノードを文字列に変換する。"
  :type)

(defmethod to-string :literal [node]
  (str (:value node)))

(defmethod to-string :variable [node]
  (str (:name node)))

(defmethod to-string :add [node]
  (str "(" (to-string (:left node)) " + " (to-string (:right node)) ")"))

(defmethod to-string :subtract [node]
  (str "(" (to-string (:left node)) " - " (to-string (:right node)) ")"))

(defmethod to-string :multiply [node]
  (str "(" (to-string (:left node)) " * " (to-string (:right node)) ")"))

(defmethod to-string :divide [node]
  (str "(" (to-string (:left node)) " / " (to-string (:right node)) ")"))

;; --- 簡易パーサー（逆ポーランド記法） ---

(defn parse-rpn
  "逆ポーランド記法の文字列を AST に変換する。"
  [expression]
  (let [tokens (str/split expression #"\s+")
        ops    {"+" add "-" subtract "*" multiply "/" divide}]
    (reduce (fn [stack token]
              (if-let [op (get ops token)]
                (let [right (peek stack)
                      stack' (pop stack)
                      left  (peek stack')
                      stack'' (pop stack')]
                  (conj stack'' (op left right)))
                (conj stack (if (re-matches #"-?\d+(\.\d+)?" token)
                              (literal (read-string token))
                              (variable (keyword token))))))
            []
            tokens)
    ))
